src/Simplex_tree/include/gudhi/Simplex_tree/multi_filtrations/Finitely_critical_filtrations.h

```diff
@@ -0,0 +1,1051 @@
+  * if it is useful (according to the method `dominate`)
+  */
+  inline void add_point(const One_critical_filtration<T> &x) {
+    const bool verbose = false;
+    if (multi_filtration_.empty()) {
+      if constexpr (verbose)
+        std::cout << "Adding x=" << x << " (currently empty)" << std::endl;
+      multi_filtration_.push_back(x);
+      return;
+    }
+    for (const auto &y : multi_filtration_){
+      if (dominates(x,y)){
+        return;
+      }
+    }
+    if constexpr (verbose)
+      std::cout << "x: " << x << " is useful, removing unnecessary entries"
+                << std::endl;
+    multi_filtration_.erase(
+        std::remove_if(multi_filtration_.begin(), multi_filtration_.end(),
+                       [&x](const One_critical_filtration<T> &y) {
+                         if constexpr (verbose) {
+                           if (dominates(y, x)) {
+                             std::cout << "Removing y=" << y << std::endl;
+                           } else {
+                             std::cout << "Keeping y=" << y << std::endl;
+                           }
+                         }
+                         return dominates(y, x);
+                       }),
+        multi_filtration_.end());
+
+    multi_filtration_.push_back(x);
+  }
+  inline void re_clean(){
+    // Ensures all points are useful again. Can be useful if points are added manually.
+    // TODO : maybe optimize
+    Multi_critical_filtration<value_type> out; // should be inf
+    out.multi_filtration_.reserve(this->multi_filtration_.size());
+    for (const auto& x : multi_filtration_){
+      out.add_point(x);
+    }
+    std::swap(multi_filtration_, out);
+  }
+
+  // easy debug
+  inline friend std::ostream &operator<<(std::ostream &stream,
+                                         const Multi_critical_filtration<T,co> &truc) {
+    if (truc.is_inf()) {
+      stream << "[inf, ..., inf]";
+      return stream;
+    }
+    if (truc.is_minus_inf()) {
+      stream << "[-inf, ..., -inf]";
+      return stream;
+    }
+    if (truc.is_nan()) {
+      stream << "[NaN]";
+      return stream;
+    }
+    stream << "(k=" << truc.multi_filtration_.size() << ")[";
+    for (const auto &machin : truc) {
+      stream << machin << "; ";
+    }
+    if (truc.multi_filtration_.size() > 0) {
+      stream << "\b"
+             << "\b";
+    }
+    stream << "]";
+    return stream;
+  }
+  inline void clear() { multi_filtration_.clear(); }
+  inline bool empty() const { return multi_filtration_.empty(); }
+  inline const OneCritical& operator[] (std::size_t i) const { return multi_filtration_[i]; }
+
+  inline typename std::vector<One_critical_filtration<T>>::iterator
+  begin() {
+    return multi_filtration_.begin();
+  }
+  inline typename std::vector<One_critical_filtration<T>>::iterator
+  end() {
+    return multi_filtration_.end();
+  }
+  inline typename std::vector<
+      One_critical_filtration<T>>::const_iterator
+  begin() const {
```

Comment:
Note that the following is equivalent (but your code is fine if you like being explicit)
```suggestion
  auto begin() const {
```